### PR TITLE
feat: add retries for FSMs

### PR DESCRIPTION
**Pull request by matt2e**

### CODE REVIEW

File: backend/controller/sql/queries.sql [hunk 1]
```diff
@@ -590,7 +607,7 @@ RETURNING true;
 -- name: FailFSMInstance :one
 UPDATE fsm_instances
 SET
-  current_state = '',
+  current_state = NULL,
```
Comment (matt2e) [inline backend/controller/sql/queries.sql:610]: `''` is not able to be read back into a option.Option[schema.Ref]`, we need to set it to null

File: backend/schema/metadataretry.go [hunk 2]
```diff
@@ -14,6 +14,7 @@ import (
 )
 
 const (
+	DefaultRetryCount  = 100
```
Comment (matt2e) [inline backend/schema/metadataretry.go:17]: This is the number of retries we attempt if the retry directive does not explicitly give a count.
What's a reasonable default?

File: integration/integration_test.go [hunk 3]
```diff
@@ -358,3 +361,88 @@ func TestFSM(t *testing.T) {
...
+		getFSMInstances(func(rows []sql.FsmInstance) error {
+			// make sure both FSM instances reached error state
+			assert.Equal(t, 2, len(rows), "unexpected number of FSM instances: %v", rows)
+			for _, r := range rows {
+				assert.Equal(t, sql.FsmStatusFailed, r.Status, "unexpected status for FSM instance %s: %v", r.Key, r.Status)
+			}
+			return nil
+		}),
+
+		getAsyncCalls(func(rows []asyncCallRow) error {
```
Comment (alecthomas) [inline integration/integration_test.go:391]: Same here, can we not just query for exact rows using `query()`?

**Review by alecthomas — COMMENTED** [review]
Awesome.

File: integration/integration_test.go [hunk 4]
```diff
@@ -358,3 +361,88 @@ func TestFSM(t *testing.T) {
...
+
+		// transition the FSM, should fail each time.
+		call("fsmretry", "startTransitionToTwo", obj{"id": "1"}, func(response obj) error { return nil }),
+		call("fsmretry", "startTransitionToThree", obj{"id": "2"}, func(response obj) error { return nil }),
+
+		sleep(8*time.Second), //6s is longest run of retries
+
+		getFSMInstances(func(rows []sql.FsmInstance) error {
+			// make sure both FSM instances reached error state
+			assert.Equal(t, 2, len(rows), "unexpected number of FSM instances: %v", rows)
```
[the remark this answers is not in the document]
Comment (matt2e) [reply integration/integration_test.go:384]: mmmmm thats why i kept having the kill ftl after failed tests 🤦‍♂️

File: backend/schema/metadataretry.go (hunk 2 above)
Comment (alecthomas) [reply backend/schema/metadataretry.go:17]: Great question. Maybe there isn't and we should be explicit, but let's start with this.

File: integration/integration_test.go [hunk 5]
```diff
@@ -358,3 +361,88 @@ func TestFSM(t *testing.T) {
...
+
+		sleep(2*time.Second),
+
+		// transition the FSM, should fail each time.
+		call("fsmretry", "startTransitionToTwo", obj{"id": "1"}, func(response obj) error { return nil }),
+		call("fsmretry", "startTransitionToThree", obj{"id": "2"}, func(response obj) error { return nil }),
+
+		sleep(8*time.Second), //6s is longest run of retries
+
+		getFSMInstances(func(rows []sql.FsmInstance) error {
```
[the remark this answers is not in the document]
Comment (matt2e) [reply integration/integration_test.go]: Cleaned up the integration test and added a `getRow` function which is like `queryRow` except it just returns the column values. This was needed so I could compare the scheduled times between rows.

File: backend/controller/controller.go [hunk 6]
```diff
@@ -780,21 +782,73 @@ func (s *Service) SendFSMEvent(ctx context.Context, req *connect.Request[ftlv1.S
...
+
+	err = tx.StartFSMTransition(ctx, instance.FSM, instance.Key, destinationRef.ToRefKey(), msg.Body, retryCount, minBackoff, maxBackoff)
 	if err != nil {
 		return nil, connect.NewError(connect.CodeInternal, fmt.Errorf("could not start fsm transition: %w", err))
 	}
 	return connect.NewResponse(&ftlv1.SendFSMEventResponse{}), nil
 }
 
+// retryParamsForTransition finds the retry metadata for the given transition and returns the retry count, min and max backoff.
+func (s *Service) retryParamsForTransition(fsm *schema.FSM, verb *schema.Verb) (count int, minBackoff, maxBackoff time.Duration, err error) {
```
Comment (alecthomas) [inline backend/controller/controller.go:806]: Maybe we should add this as a method on `schema.FSM`?

File: backend/controller/dal/async_calls.go [hunk 7]
```diff
@@ -114,9 +124,27 @@ func (d *DAL) CompleteAsyncCall(ctx context.Context, call *AsyncCall, result eit
...
 
 	case either.Right[[]byte, string]: // Failure message.
-		_, err = tx.db.FailAsyncCall(ctx, result.Get(), call.ID)
-		if err != nil {
-			return translatePGError(err)
+		if call.RemainingAttempts > 0 {
+			nextBackoff := call.Backoff * 2
+			if nextBackoff > call.MaxBackoff {
+				nextBackoff = call.MaxBackoff
+			}
```
Comment (alecthomas) [inline backend/controller/dal/async_calls.go:131]:
```suggestion
			nextBackoff := min(call.Backoff * 2, call.MaxBackoff)
```

File: backend/controller/sql/queries.sql [hunk 8]
```diff
@@ -530,6 +534,19 @@ SET
...
 
+-- name: FailAsyncCallWithRetry :one
+WITH updated AS (
+  UPDATE async_calls
+  SET state = 'error'::async_call_state,
+      error = @error::TEXT
+  WHERE id = @id::BIGINT
+  RETURNING *
+)
+INSERT INTO async_calls (verb, origin, request, remaining_attempts, backoff, max_backoff, scheduled_at)
```
Comment (alecthomas) [inline backend/controller/sql/queries.sql:545]: Nice

**Review by alecthomas — APPROVED** [review]
Nice this is awesome.

File: backend/controller/controller.go (hunk 6 above)
Comment (matt2e) [reply backend/controller/controller.go]: Tidied that up by also removing the specialized getters for `MinBackoffDuration`/`MaxBackoffDuration` and instead returning a normalized version with `RetryParams` https://github.com/TBD54566975/ftl/pull/1583/commits/0c85daad74839c5207018aafbdd5accf9aa5cc71

File: backend/controller/dal/async_calls.go (hunk 7 above)
Comment (matt2e) [reply backend/controller/dal/async_calls.go:131]: copilot told me there was no such built in thing in go 😤

**Pull request merged**